pointpats/spacetime.py

```diff
@@ -629,3 +658,1008 @@ def _shuffle_matrix(X, ids):
+        )
+        blues = g[g.color == colors["neighbor"]]
+        if blues.shape[0] == 0:
+            warn("empty neighbor set.")
+        else:
+            m = blues.explore(m=m, color=colors["neighbor"], style_kwds=style_kwds)
+        m = g[g.color == colors["focal"]].explore(
+            m=m, color=colors["focal"], style_kwds=style_kwds
+        )
+
+        if plot_edges:
+
+            # edges between hotspot and st-neighbors
+            g = g.set_index("index")
+            ghs = self.hotspots(crit=crit, inference=inference)
+            ghs = ghs.dropna()
+            origins = g.loc[ghs.focal].geometry
+            destinations = g.loc[ghs.neighbor].geometry
+            ods = zip(origins, destinations)
+            lines = gpd.GeoSeries([LineString(od) for od in ods])
+            lines.crs = g.crs
+            lines.explore(m=m, color=edge_color, style_kwds={"weight": edge_weight})
+
+        return m
+
+    def _gdfhs(self, crit=0.05, inference="permutation"):
+        # merge df with self.hotspots
+        return gpd.GeoDataFrame(
+            self._gdf.merge(
+                self.hotspots(crit=crit, inference=inference),
+                left_index=True,
+                right_on="focal",
+            )
+        )
+
+
+def _spacetime_points_to_arrays(dataframe, time_col):
+    """convert long-form geodataframe into arrays for kdtree
+
+    Parameters
+    ----------
+    dataframe : geopandas.GeoDataFrame
+        geodataframe with point geometries
+    time_col : str
+        name of the column on dataframe that stores time values
+
+    Returns
+    -------
+    tuple
+        two numpy arrays holding spatial coodinates s_coords (n,2)
+        and temporal coordinates t_coords (n,1)
+
+    """
+    if dataframe.crs is None:
+        warn(
+            "There is no CRS set on the dataframe. The KDTree will assume coordinates "
+            "are stored in Euclidean distances"
+        )
+    else:
+        if dataframe.crs.is_geographic:
+            raise ValueError(
+                "The input dataframe must be in a projected coordinate system."
+            )
+
+    assert dataframe.geom_type.unique().tolist() == [
+        "Point"
+    ], "The Knox statistic is only defined for Point geometries"
+
+    # kdtree wont operate on datetime
+    if is_numeric_dtype(dataframe[time_col].dtype) is False:
+        raise ValueError(
+            "The time values must be stored as "
+            f"a numeric dtype but the column {time_col} is stored as "
+            f"{dataframe[time_col].dtype}"
+        )
+
+    s_coords = np.vstack((dataframe.geometry.x.values, dataframe.geometry.y.values)).T
+    t_coords = np.vstack(dataframe[time_col].values)
+
+    return s_coords, t_coords
+
+
+def _time_to_int(df, column, fmt="%Y%m%d"):
+    """Create an integer valued column for time"""
+    time_series = pandas.to_datetime(df[column], format=fmt)
+    min_time = time_series.min()
+    time_int = time_series.appy(lambda x: x - min_time)
```

Comment:
Here's a typo in `apply`. I also don't think you need apply here.
```suggestion
    time_int = time_series - min_time
```

Reply:
we were going to try an introspect any datetime col if passed, but then punted and decided the time col needs to be given as integer from the user.

so this is just a vestige of that idea and can be removed